core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java

```diff
@@ -0,0 +1,361 @@
+        .withOperandSupplier(b0 ->
+            b0.operand(Join.class).inputs(
+                b1 -> b1.operand(RelNode.class).anyInputs(),
+                b2 -> b2.operand(Values.class).predicate(Values::isSingleValue).noInputs()))
+        .withDescription("PruneJoinSingleValue(right)");
+
+    @Override default SingleValueOptimizationRules.PruneSingleValueRule toRule() {
+      return new SingleValueOptimizationRules.PruneSingleValueRule(this) {
+        @Override public void onMatch(RelOptRuleCall call) {
+          Predicate<Join> predicate = elibilityPredicate(false);
+          final Join join = call.rel(0);
+          final Values values = call.rel(2);
+          final RelNode right = call.rel(1);
+          final RelBuilder relBuilder = call.builder();
+          RelNode transformed =
+              new SingleValueRelTransformer(join,
+                  values.tuples.get(0).stream().collect(Collectors.toList()), right,
+                  predicate, false, getRexTransformer(relBuilder.getRexBuilder(),
+                  join.getJoinType())).transform(relBuilder);
+          if (transformed != null) {
+            call.transformTo(transformed);
+          }
+        }
+      };
+    }
+  }
+
+  /** Configuration for a rule that simplifies join node with constant row on its left input. */
+  @Value.Immutable
+  interface JoinLeftSingleRuleConfig extends PruneSingleValueRule.Config {
+    JoinLeftSingleRuleConfig DEFAULT = ImmutableJoinLeftSingleRuleConfig.of()
+        .withOperandSupplier(b0 ->
+            b0.operand(Join.class).inputs(
+                b1 -> b1.operand(Values.class).predicate(Values::isSingleValue).noInputs(),
+                b2 -> b2.operand(RelNode.class).anyInputs()))
+        .withDescription("PruneJoinSingleValueRule(left)");
+
+    @Override default SingleValueOptimizationRules.PruneSingleValueRule toRule() {
+      return new SingleValueOptimizationRules.PruneSingleValueRule(this) {
+        @Override public void onMatch(RelOptRuleCall call) {
+          Predicate<Join> predicate = elibilityPredicate(true);
+          final Join join = call.rel(0);
+          final Values values = call.rel(1);
+          final RelNode right = call.rel(2);
+          final RelBuilder relBuilder = call.builder();
+          RelNode transformed =
+              new SingleValueRelTransformer(join,
+                  values.tuples.get(0).stream().collect(Collectors.toList()), right,
+                  predicate, true, getRexTransformer(relBuilder.getRexBuilder(),
+                  join.getJoinType())).transform(relBuilder);
+          if (transformed != null) {
+            call.transformTo(transformed);
+          }
+        }
+      };
+    }
+  }
+
+  /** Configuration for a rule that simplifies join node with a project on a constant row
+   *  on its left input. */
+  @Value.Immutable
+  interface JoinLeftSingleValueRuleWithExprConfig extends PruneSingleValueRule.Config {
```

Comment:
`JoinLeftSingleValueRuleWithExprConfig` and `JoinRightSingleValueRuleWithExprConfig` has many same code.

Can we extract them?

Reply:
Agree.

Reply:
Thanks @JiajunBernoulli  and @macroguo-ghy  for the suggestion. I moved the redundant code to onMatch base class function.